File: sdks/rust/examples/performance_monitoring.rs
//! Performance monitoring example for Nexus Rust SDK

use nexus_sdk::*;

#[tokio::main]
async fn main() -> Result<()> {
    let client = NexusClient::new("http://localhost:15474")?;

    tracing::info!("=== Performance Monitoring Example ===\n");

    // Get query statistics
    tracing::info!("1. Getting query statistics...");
    match client.get_query_statistics().await {
        Ok(stats) => {
            tracing::info!("   Total queries: {}", stats.statistics.total_queries);
            tracing::info!("   Successful: {}", stats.statistics.successful_queries);
            tracing::info!("   Failed: {}", stats.statistics.failed_queries);
            tracing::info!(
                "   Average execution time: {}ms",
                stats.statistics.average_execution_time_ms
            );
            tracing::info!("   Slow queries: {}", stats.statistics.slow_query_count);
            tracing::info!("   Patterns tracked: {}", stats.patterns.len());
        }
        Err(e) => tracing::info!("   Error: {}", e),
    }

    // Get slow queries
    tracing::info!("\n2. Getting slow queries...");
    match client.get_slow_queries().await {
        Ok(slow_queries) => {
            tracing::info!("   Found {} slow queries", slow_queries.count);
            for (i, query) in slow_queries.queries.iter().take(5).enumerate() {
                tracing::info!(
                    "   {}. {}ms - {}",
                    i + 1,
                    query.execution_time_ms,
                    query.query
                );
            }
        }
        Err(e) => tracing::info!("   Error: {}", e),
    }

    // Get plan cache statistics
    tracing::info!("\n3. Getting plan cache statistics...");
    match client.get_plan_cache_statistics().await {
        Ok(cache_stats) => {
            tracing::info!("   Cached plans: {}", cache_stats.cached_plans);
            tracing::info!("   Max size: {}", cache_stats.max_size);
            tracing::info!(
                "   Current memory: {} bytes",
                cache_stats.current_memory_bytes
            );
            tracing::info!("   Max memory: {} bytes", cache_stats.max_memory_bytes);
            tracing::info!("   Hit rate: {:.2}%", cache_stats.hit_rate * 100.0);
        }
        Err(e) => tracing::info!("   Error: {}", e),
    }

    // Clear plan cache
    tracing::info!("\n4. Clearing plan cache...");
    match client.clear_plan_cache().await {
        Ok(_) => tracing::info!("   Plan cache cleared successfully"),
        Err(e) => tracing::info!("   Error: {}", e),
    }

    tracing::info!("\n=== Example completed ===");
    Ok(())
}

File: sdks/python/examples/performance_monitoring.py
"""Performance monitoring example for Nexus Python SDK."""

import asyncio
from nexus_sdk import NexusClient


async def main():
    """Run performance monitoring example."""
    async with NexusClient("http://localhost:15474") as client:
        print("=== Performance Monitoring ===\n")

        # Get query statistics
        print("1. Getting query statistics...")
        try:
            stats = await client.get_query_statistics()
            print(f"   Total queries: {stats.statistics.total_queries}")
            print(f"   Successful: {stats.statistics.successful_queries}")
            print(f"   Failed: {stats.statistics.failed_queries}")
            print(f"   Average time: {stats.statistics.average_execution_time_ms}ms\n")
        except Exception as e:
            print(f"   Error: {e}\n")

        # Get slow queries
        print("2. Getting slow queries...")
        try:
            slow_queries = await client.get_slow_queries()
            print(f"   Found {slow_queries.count} slow queries")
            for query in slow_queries.queries[:5]:  # Show first 5
                print(f"   - {query.query[:50]}... ({query.execution_time_ms}ms)\n")
        except Exception as e:
            print(f"   Error: {e}\n")

        # Get plan cache statistics
        print("3. Getting plan cache statistics...")
        try:
            cache_stats = await client.get_plan_cache_statistics()
            print(f"   Cached plans: {cache_stats.cached_plans}")
            print(f"   Hit rate: {cache_stats.hit_rate:.2%}")
            print(f"   Memory usage: {cache_stats.current_memory_bytes} bytes\n")
        except Exception as e:
            print(f"   Error: {e}\n")

        # Clear plan cache
        print("4. Clearing plan cache...")
        try:
            result = await client.clear_plan_cache()
            print(f"   Cache cleared: {result}\n")
        except Exception as e:
            print(f"   Error: {e}\n")


if __name__ == "__main__":
    asyncio.run(main())

